main.py

```diff
@@ -569,15 +569,16 @@ def on_link_liveness_disabled(self, event) -> None:
 
     def get_links_from_interfaces(self, interfaces) -> dict:
         """Get links from interfaces."""
-        links = {}
+        links_found = {}
+        links = self.links.copy()
         for interface in interfaces:
-            for link in self.links.values():
+            for link in links.values():
                 if any((
                     interface.id == link.endpoint_a.id,
                     interface.id == link.endpoint_b.id,
                 )):
-                    links[link.id] = link
-        return links
+                    links_found[link.id] = link
+        return links_found
```

Comment:
Why the copy here? The original code iterates over `self.links` and updates `links`, so it is not changing the dictionary being iterated. I'd just rename `links` to `links_found`, the name looks more descriptive.

Reply:
Hi, @ajoaoff, in this case self.links is a shared dict memory [being iterated on](https://github.com/kytos-ng/topology/blob/master/main.py#L575-L579), and `self.links` can get new links inserted [from other threads](https://github.com/kytos-ng/topology/blob/master/main.py#L843-L855), resulting in the same problem. So, either we add a shallow copy here or protect at the call site with [`self.link_locks`](https://github.com/kytos-ng/topology/blob/master/main.py#L44), but then when there are too many caller it start to get cumbersome putting locks everywhere (or we put the lock inside the method facilitating), since we don't have that many callers, putting a lock here wouldn't be that bad, so I'll put a lock instead, the rest of he code is already using this pattern, and then we can reassess in the future. Having locks also can lead to other issues but these parts here shouldn't lead to similar slowness that we've seen.

Reply:
I've pushed the change here https://github.com/kytos-ng/topology/pull/109/commits/8efd02bdbe7a8c69668db94a375aa527196f033b